File: src/platform/memory.rs
use core::mem::size_of;
use core::num::NonZeroUsize;
use core::slice::from_raw_parts_mut;

pub fn map<T: Sized>(address: usize, amount: usize) -> Option<&'static mut[T]> {
    unsafe {
        self::memory
            ::map(address, size_of::<T>() * amount)
            .map(|ptr| from_raw_parts_mut(ptr.get() as *mut _, amount))
    }
}

pub fn unmap<T: Sized>(memory: &[T]) -> Option<()> {
    unsafe { 
        self::memory::unmap(
            memory.as_ptr() as usize,
            memory.len() * size_of::<T>()
        )
    }
}

pub fn commit<T: Sized>(memory: &[T]) -> Option<()> {
    unsafe {
        self::memory::commit(
            memory.as_ptr() as usize,
            memory.len() * size_of::<T>()
        )
    }
}

pub fn decommit<T: Sized>(memory: &[T]) -> Option<()> {
    unsafe {
        self::memory::decommit(
            memory.as_ptr() as usize,
            memory.len() * size_of::<T>()
        )
    }
}

#[cfg(unix)]
mod memory {
    use libc::*;
    use super::NonZeroUsize;

    pub unsafe fn map(address: usize, bytes: usize) -> Option<NonZeroUsize> {
        let prot = PROT_READ | PROT_WRITE;
        let flags = MAP_FIXED | MAP_PRIVATE | MAP_ANONYMOUS;
        match mmap(address as *mut _, bytes, prot, flags, -1, 0) as usize {
            ptr if ptr != address => None,
            ptr => Some(NonZeroUsize::new_unchecked(ptr)),
        }
    }

    pub unsafe fn unmap(address: usize, bytes: usize) -> Option<()> {
        match munmap(address as *mut _, bytes) {
            0 => None,
            _ => Some(()),
        }
    }

    pub unsafe fn commit(address: usize, bytes: usize) -> Option<()> {
        // linux over-commits by default
        Some(())
    }

    pub unsafe fn decommit(address: usize, bytes: usize) -> Option<()> {
        match madvise(address as *mut _, bytes, MADV_DONTNEED) {
            0 => Some(()),
            _ => None,
        }
    }
}

#[cfg(windows)]
mod memory {
    use super::NonZeroUsize;
    use winapi::shared::minwindef::*;
    use winapi::um::{winnt::*, memoryapi::*};

    pub unsafe fn map(address: usize, bytes: usize) -> Option<NonZeroUsize> {
        match VirtualAlloc(address as *mut _, bytes, MEM_RESERVE, PAGE_READWRITE) as usize {
            0 => None,
            ptr => Some(NonZeroUsize::new_unchecked(ptr)),
        }
    }

    pub unsafe fn unmap(address: usize, bytes: usize) -> Option<()> {
        match VirtualFree(address as *mut _, 0, MEM_RELEASE) {
            TRUE => Some(()),
            _ => None,
        }
    }

    pub unsafe fn commit(address: usize, bytes: usize) -> Option<()> {
        match VirtualAlloc(address as *mut _, bytes, MEM_COMMIT, PAGE_READWRITE) {
            ptr if ptr.is_null() => None,
            _ => Some(()),
        }
    }

    pub unsafe fn decommit(address: usize, bytes: usize) -> Option<()> {
        match VirtualFree(address as *mut _, bytes, MEM_DECOMMIT) {
            TRUE => Some(()),
            _ => None,
        }
    }
}

File: src/platform/memory.zig
const std = @import("std");
const builtin = @import("builtin");

const linux = std.os.linux;
const windows = builtin.os == builtin.Os.windows;

const MADV_DONTNEED = 4;
const MEM_COMMIT = 0x1000;
const MEM_RESERVE = 0x2000;
const MEM_DECOMMIT = 0x4000;
const MEM_RELEASE = 0x8000;
const PAGE_READWRITE = 0x04;
const INVALID = std.math.maxInt(usize);

extern "kernel32" stdcallcc fn VirtualFree(address: usize, bytes: usize, free_type: u32) bool;
extern "kernel32" stdcallcc fn VirtualAlloc(address: usize, bytes: usize, alloc_type: u32, protect: u32) usize;

pub fn map(address: usize, bytes: usize) usize {
    if (windows)
        return VirtualAlloc(address, bytes, MEM_RESERVE, PAGE_READWRITE);

    const prot = linux.PROT_READ | linux.PROT_WRITE;
    const flags = linux.MAP_FIXED | linux.MAP_PRIVATE | linux.MAP_ANONYMOUS;
    const addr = linux.syscall6(linux.SYS_mmap, address, bytes, prot, flags, INVALID, 0);

    return if (addr == address) address else 0;
}

pub fn unmap(address: usize, bytes: usize) bool {
    if (windows)
        return VirtualFree(address, 0, MEM_RELEASE);
    return linux.syscall2(linux.SYS_munmap, address, bytes) == 0;
}

pub fn commit(address: usize, bytes: usize) bool {
    if (windows)
        return VirtualAlloc(address, bytes, MEM_COMMIT, PAGE_READWRITE) == address;
    return true; // linux overcommits by default
}

pub fn decommit(address: usize, bytes: usize) bool {
    if (windows)
        return VirtualFree(address, bytes, MEM_DECOMMIT);
    return linux.syscall3(linux.SYS_madvise, address, bytes, MADV_DONTNEED) == 0;
}

pub inline fn ptrCast(comptime To: type, from: var) To {
    return @intToPtr(To, @ptrToInt(from));
}

pub fn aligned(value: usize, comptime alignment: usize) usize {
    return value + (alignment - (value % alignment));
}

pub fn nextPow2(value: u32) u32 {
    const shift: u32 = 32 - @clz(value - 1);
    return u32(1) << @truncate(u5, shift);
}

pub fn ptrToSlice(comptime T: type, ptr: [*]T, size: usize) []T {
    // WARNING: zig hack
    // this relies on the structure layout of a slice since theres no way to unsafe-ly create one
    return @bitCast([]T, [2]usize { @ptrToInt(ptr), size });
}

test "memory mapping" {
    const assert = @import("std").debug.assert;
    
    const memory = @intToPtr([*]u8, map(1 << 32, 4096));
    defer assert(unmap(1 << 32, 4096));
    
    assert(commit(@ptrToInt(memory), 4096));
    defer assert(decommit(@ptrToInt(memory), 4096));

    memory[0] = 0xff;
    assert(memory[0] == 0xff);
}